### PR TITLE
Cs/sc 1821 add to script for sizing model params

### PR DESCRIPTION
**Pull request by Charl1996**
## Product Description
No UI changes.

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-1821)

This PR speaks to adding additional calculations that can be used to give insights into the CommCare Resource Model parameters for the config file. Some additional calculations was added which might assist in determining the input values to the resource model usage section as well as some additional parameters such as synclogs and case transactions (the list is not exhaustive, so feel free to suggest adding more).

Changes:
I added an object `ResourceModel` which allows keeping state in order to reuse already-calculated values and also allows for outputting the values in a separate phase to collecting the values. 

## Feature Flag
None

### Safety story
Local testing

### Automated test coverage
No automated tests needed

### QA Plan
No QA needed

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 1]
```diff
@@ -179,6 +239,12 @@ def _print_table(self, headers, rows):
         SimpleTableWriter(self.stdout, row_formatter).write_table(headers, rows)
         self.stdout.write('')
 
+    def _print_section_title(self, title_string):
+        self.stdout.write('')
+        # self.stdout.write('=' * len(title_string))
```
Comment (kaapstorm) [inline corehq/apps/reports/management/commands/project_stats_report.py:244]: Did you miss this one?

**Review by kaapstorm — APPROVED** [review]

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 2]
```diff
@@ -45,6 +45,33 @@
...
+        'case_transactions_total': None,
+        'case_indices': None,
+        'synclogs_monthly': None,
+    }
+
+    @classmethod
+    def set_stat(cls, name, value):
+        if name not in cls.stats.keys():
+            raise Exception(f'{name} is not recognized')
+        cls.stats[name] = value
```
[the remark this answers is not in the document]
Comment (kaapstorm) [reply corehq/apps/reports/management/commands/project_stats_report.py:93]: Oh, P.S. `ValueError` might be more helpful than `Exception`.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 3]
```diff
@@ -82,18 +136,23 @@ def handle(self, domain, months, csv, **options):
             .domain(domain).values_list("_id", flat=True)
         )
 
+        self._doc_counts()
         self._forms_per_user_per_month()
         self._cases_created_per_user_per_month()
         self._cases_updated_per_user_per_month()
+        self._case_transactions()
+        self._case_indices()
+        self._synclogs()
```
Comment (sravfeyn) [inline corehq/apps/reports/management/commands/project_stats_report.py:145]: Shouldn't this include `DeviceReportEntry`?

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 4]
```diff
@@ -82,18 +136,23 @@ def handle(self, domain, months, csv, **options):
             .domain(domain).values_list("_id", flat=True)
         )
 
+        self._doc_counts()
```
Comment (sravfeyn) [inline corehq/apps/reports/management/commands/project_stats_report.py:139]: These methods are setting stats, but they are simply called with a noun, so it might not be easy for someone to understand what the method is doing without reading the method. I would rename them to `set_<stat_name>` or something on those lines. Also `_` at the beginning might not be necessary in this case.

Just a stylistic comment, nothing major.

**Comment by sravfeyn** [comment]
Looks good, left a couple minor comments.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 5]
```diff
@@ -45,6 +45,60 @@
 )
 
 
+class ResourceModel(object):
+
+    TOTAL_USERS = 'total_users'
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:50]: This is _much_ cleaner!

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 6]
```diff
@@ -295,20 +361,15 @@ def _attachment_sizes(self):
...
+            ResourceModel.set_stat(ResourceModel.ATTACHMENTS, result)
 
     def _ucr(self):
         static_datasources = StaticDataSourceConfiguration.by_domain(self.domain)
         dynamic_datasources = DataSourceConfiguration.by_domain(self.domain)
-        self._print_value('Static UCR data sources', len(static_datasources))
-        self._print_value('Dynamic UCR data sources', len(dynamic_datasources))
+
+        ResourceModel.set_stat(ResourceModel.STATIC_DATASOURCES, len(static_datasources))
+        ResourceModel.set_stat(ResourceModel.DYNAMIC_DATASOURCES, len(dynamic_datasources))
```
Comment (proteusvacuum) [inline corehq/apps/reports/management/commands/project_stats_report.py:372]: I know you didn't write this, but this will pull all the configs into memory. You could instead use the reduced couch query with something like: 
```py
        DataSourceConfiguration.view(
            'userreports/data_sources_by_build_info',
            startkey=key,
            endkey=key + [{}],
            reduce=True,
            include_docs=False
        )
```
since the reduce function for that view is `_count`.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 7]
```diff
@@ -339,8 +400,200 @@ def _get_table_size(config):
             for datasource in static_datasources + dynamic_datasources
         ], key=lambda r: r[-1] if r[-1] != 'Table not found' else 0)
 
+        ResourceModel.set_stat(ResourceModel.DATASOURCES_INFO, rows)
+
+    def _case_transactions(self):
+        db_name = get_db_aliases_for_partitioned_query()[0]
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:406]: @proteusvacuum That is a great question! I saw that this command (`project_stats_report`) already uses only one partition to gather data, so I queried on the #gtd-tech-team slack channel about it, upon which Simon replied
> For this command specifically we are just producing estimate numbers and are assuming that data is 'evenly' distributed between the shards 

Back then I just went with this assumption, but maybe it's time to challenge it? 

Since the resource model will only really rely on the biggest contributors to the resource usage, the question I guess becomes how accurate we want those contributions estimated. 

I'm wondering now if we say, "assuming data is evenly distributed", how are we (or are we?) taking the number of shards into account. Querying one shard with evenly distributed data among 10 shards yields 1/10 of the total data.

@snopoke Any thoughts?

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 8]
```diff
@@ -339,8 +400,200 @@ def _get_table_size(config):
...
+    def _synclogs(self):
+        db_name = get_db_aliases_for_partitioned_query()[0]
+        db_cursor = connections[db_name].cursor()
+
+        with db_cursor as cursor:
+            cursor.execute("""
+                SELECT COUNT(*), d.count
+                FROM (
+                    SELECT COUNT(*) AS count
+                    FROM phone_synclogsql
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:471]: Mmm...no...I was not aware of the roll over, but I see the problem. But I'm actually thinking now...it seems there's a [configuration](https://github.com/dimagi/commcare-cloud/blob/a50ada52c88d22937577f088551bc1aac751a848/docs/commcare-cloud/env/postgresql_yml.md#synclogs) that allows a user to specify a different table to store synclogs. If I look at [this](https://github.com/dimagi/commcare-hq/blob/80fb62fe7bf9731f7cb6a614e8753a0b18e8983b/corehq/ex-submodules/casexml/apps/phone/tasks.py#L77) it seems there's maybe a different way I should approach the synclogs...maybe use something like `db = router.db_for_read(SyncLogSQL)`...

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 9]
```diff
@@ -45,6 +45,60 @@
...
+        CASE_TRANSACTIONS_TOTAL: None,
+        CASE_INDICES: None,
+        SYNCLOGS_MONTHLY_FACTOR: None,
+        STATIC_DATASOURCES: None,
+        DYNAMIC_DATASOURCES: None,
+        DATASOURCES_INFO: None,
+    }
+
+    @classmethod
+    def set_stat(cls, name, value):
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:90]: Actually...after seeing the stackoverflow implementation I kinda like it. It feels a bit more compact.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 10]
```diff
@@ -339,8 +400,200 @@ def _get_table_size(config):
...
+            """, [self.domain])
+            result = cursor.fetchall()
+
+            total_user_synclogs = 0
+            total_users = 0
+            for num_users, num_synclogs in result:
+                total_users += num_users
+                total_user_synclogs += num_users * num_synclogs
+
+            ResourceModel.set_stat(ResourceModel.SYNCLOGS_MONTHLY_FACTOR, total_user_synclogs / total_users)
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:486]: @proteusvacuum
> how this becomes a "monthly" value

You're actually right, it doesn't! It shows the synclogs per user (and it's supposed to). 

(...then why did I output it as per month?... this seems like a rather careless mistake on my part)

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 11]
```diff
@@ -339,8 +400,200 @@ def _get_table_size(config):
...
+        self.stdout.write('')
         self.stdout.write('UCR datasource sizes')
         self._print_table(
             ['Datasource name', 'Row count (approximate)', 'Doc type', 'Size (bytes)'],
             rows
         )
+
+    def _output_case_transactions(self):
+        case_transactions = ResourceModel.get_stat(ResourceModel.CASE_TRANSACTIONS_FACTOR)
+        self.stdout.write(f'Average cases updated per form per month: {case_transactions}')
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:592]: Same as [here](https://github.com/dimagi/commcare-hq/pull/30837/files#r828812808).

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 6 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:372]: That could work, yes, but we are not interested only in the count. The actual data is used also [here](https://github.com/dimagi/commcare-hq/blob/c7c5ea9809f2af0808dbbe350e9ebda3875803be/corehq/apps/reports/management/commands/project_stats_report.py#L400).

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 12]
```diff
@@ -136,25 +123,26 @@ def handle(self, domain, months, csv, **options):
...
+        self.collect_devicelogs()
+        self.collect_case_to_case_index_ratio()
+        self.collect_ledgers_per_case()
+        self.collect_attachment_sizes()
+        self.collect_ucr_data()
+
+        self.output_stats()
+
+    def collect_doc_counts(self):
+        resource_model.__setitem__('forms_total', FormES().domain(self.domain).count())
```
Comment (proteusvacuum) [inline corehq/apps/reports/management/commands/project_stats_report.py:142]: I think this would now be
```py
resource_model['forms_total'] = FormES().domain(self.domain).count()
...
```
(and the same for everywhere else)

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 13]
```diff
@@ -575,10 +578,10 @@ def _output_attachment_sizes(self):
         )
 
     def _output_ucr(self):
-        self.stdout.write(f"Static UCR data sources: {ResourceModel.get_stat(ResourceModel.STATIC_DATASOURCES)}")
-        self.stdout.write(f"Dynamic UCR data sources: {ResourceModel.get_stat(ResourceModel.DYNAMIC_DATASOURCES)}")
+        self.stdout.write(f"Static UCR data sources: {resource_model.__getitem__('static_datasources')}")
```
Comment (proteusvacuum) [inline corehq/apps/reports/management/commands/project_stats_report.py:581]: This will now just be 
```py
resource_model['static_datasources']
```

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 12 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:142]: (omw, Charl... :facepalm: )

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 14]
```diff
@@ -114,31 +161,28 @@ def _forms_per_user_per_month(self):
...
+        resource_model['total_users'] = total_users
+
+        if months > 0:
+            resource_model['monthly_forms_per_user'] = total_average_forms/months
+
+        monthly_user_stats = user_stat_from_malt\
+            .filter(user_type__in=['CommCareUser'])\
+            .filter(user_id__in=self.active_not_deleted_users)\
+            .filter(num_of_forms__gte=performance_threshold)
+
```
Comment (proteusvacuum) [inline corehq/apps/reports/management/commands/project_stats_report.py:182]: nit, if you surround this in `()` you wouldn't need the `\`

```suggestion
        monthly_user_stats = (user_stat_from_malt
            .filter(user_type__in=['CommCareUser'])
            .filter(user_id__in=self.active_not_deleted_users)
            .filter(num_of_forms__gte=performance_threshold))

```

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply corehq/apps/reports/management/commands/project_stats_report.py:458]: That might have been a misleading code snippet. 
I mean that synclogs are stored in separate partitions based on the week 
https://github.com/dimagi/commcare-hq/blob/cd89af8efb71b8a178e999297bd65c8dca2ddd9b/corehq/ex-submodules/casexml/apps/phone/models.py#L425-L427
I would have thought to get the total number of synclogs, you would need to query each one of these tables. I think the alternative, as you say is to use the ORM to query it.

**Review by kaapstorm — APPROVED** [review]
I like the direction this has taken.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 15]
```diff
@@ -43,6 +43,47 @@
...
+    'case_indices',
+    'synclogs_per_user',
+    'static_datasources',
+    'dynamic_datasources',
+    'datasources_info',
+    'devicelogs_per_user',
+]
+
+
+class ResourceModel(object):
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:71]: Why not use a dataclass or attrs class?

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/reports/management/commands/project_stats_report.py:393]: You could run some queries to validate the assumption or you could run the stats query on all the shards.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 16]
```diff
@@ -339,8 +387,220 @@ def _get_table_size(config):
...
+            cursor.execute("""
+                SELECT COUNT(*) as num_forms, d.count as num_updates
+                FROM (
+                    SELECT COUNT(*) as count
+                    FROM form_processor_casetransaction
+                    WHERE case_id IN (
+                        SELECT case_id
+                        FROM form_processor_commcarecasesql
+                        WHERE domain = %s
+                    )
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:406]: This seems like an inefficient way of doing a join, can't it be re-written as:
```
SELECT COUNT(*) as count
FROM form_processor_casetransaction t JOIN form_processor_commcarecasesql c on t.case_id = c.case_id
WHERE c.domain = %s
GROUP BY form_id
```

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 17]
```diff
@@ -339,8 +387,220 @@ def _get_table_size(config):
...
+                    GROUP BY form_id
+                ) AS d
+                GROUP BY d.count;
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:409]: Grouping by a count seems odd, why not just sum the values which is what you're doing below anyway.
```
select count(*) as num_forms, sum(d.count) as num_transactions from (....) as d;
```

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 18]
```diff
@@ -339,8 +387,220 @@ def _get_table_size(config):
...
+
+            resource_model['case_transactions_per_form'] = running_form_case_updates / total_forms
+
+            cursor.execute("""
+                SELECT COUNT(*)
+                FROM form_processor_casetransaction
+                WHERE case_id IN (
+                    SELECT case_id
+                    FROM form_processor_commcarecasesql
+                    WHERE domain = %s
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:427]: another weird join

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/reports/management/commands/project_stats_report.py:458]: Since we use use table inheritance it is possible to query just the parent table which will retrieve data from the child tables in the background.

Having said that it is likely that this data will be unreliable since we delete synclog data after a certain amount of time and also after successful syncs so I don't think it's possible to get the data you want from this table.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 19]
```diff
@@ -339,8 +387,220 @@ def _get_table_size(config):
...
+        with db_cursor as cursor:
+            cursor.execute("""
+                SELECT COUNT(*), d.count
+                FROM (
+                    SELECT COUNT(*) AS count
+                    FROM phone_synclogsql
+                    WHERE domain = %s
+                    GROUP BY user_id
+                ) AS d
+                GROUP BY d.count
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:462]: similar comment to above about using `sum` instead of doing this weird group by count.

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 17 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:409]: Well spotted!

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:458]: @snopoke I presume you're referring to [this](https://github.com/dimagi/commcare-hq/blob/80fb62fe7bf9731f7cb6a614e8753a0b18e8983b/corehq/ex-submodules/casexml/apps/phone/tasks.py#L73)?

> Having said that it is likely that this data will be unreliable since we delete synclogs ... after successful syncs so I don't think it's possible to get the data you want from this table.

I guess the fact that synclogs is deleted after each successfully sync would result in the unreliability of the true number of synclogs produced by each user, but that's not the true aim here; the aim is rather the number of synclogs produced on average by the users that remain on the server to take up resources within the approx. 2 months before they are cleared.

I see we do consider the synclogs quite often as part of the resource model inputs, hence why I thought to include it as part of the output of the `project_stats_report` command...but I agree, the synclogs stat could vary vastly from month to month, but I'm not sure how else to report it?

[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/reports/management/commands/project_stats_report.py:458]: I was referring to this method that is called during form processing: https://github.com/dimagi/commcare-hq/blob/da986da1447bb23f42af0827b74c92dc1f98b844/corehq/ex-submodules/casexml/apps/case/util.py#L96

I think it's fine to include this data. It is useful in terms of DB disk but not as an indicator of sync frequency.

**Comment by Charl1996** [comment]
Major change from new commits include:
- Iterating over all DB partitions instead of only one

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 20]
```diff
@@ -82,20 +121,25 @@ def handle(self, domain, months, csv, **options):
...
+        self.collect_synclogs()
+        self.collect_devicelogs()
+        self.collect_ledgers_per_case()
+        self.collect_attachment_sizes()
+        self.collect_ucr_data()
+
+        self.output_stats()
+
+    def collect_doc_counts(self):
+        resource_model['forms_total'] = FormES().domain(self.domain).count()
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:139]: Have you reviewed the default filters that are applied to these queries? Some of them might be worth removing

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 21]
```diff
@@ -208,22 +258,22 @@ def _cases_updated_per_user_per_month(self):
...
-    def _ledgers_per_case(self):
-        db_name = get_db_aliases_for_partitioned_query()[0]  # just query one shard DB
-        results = (
-            LedgerValue.objects.using(db_name).filter(domain=self.domain)
-            .values('case_id')
-            .annotate(ledger_count=Count('pk'))
-        )[:100]
+        resource_model['monthly_user_cases_updated'] = final_stats
 
+    def collect_ledgers_per_case(self):
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:263]: should this be updating the resource_model like the other methods?

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 22]
```diff
@@ -339,8 +378,221 @@ def _get_table_size(config):
...
+                total_form_case_updates += form_case_updates
+
+                cursor.execute("""
+                    SELECT COUNT(*) as count
+                    FROM form_processor_casetransaction t
+                    JOIN form_processor_commcarecasesql c on t.case_id = c.case_id
+                    WHERE c.domain = %s
+                """, [self.domain])
+                (transactions,) = cursor.fetchone()
+                total_transactions += transactions
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:417]: how is this different from `total_form_case_updates`?

**Review by snopoke — COMMENTED** [review]
Other than the comments already made I have a few general comments:

Is it necessary to run queries on all shard databases? Could you not run on 1 and multiply by the number of shard DBs?

Some of these queries are going to be very expensive. Have they been profiled on prod?

I see that we've moved some from using `estimate_row_count` to doing actual row counts. What was the reason for that change?

The synclog numbers are going to be quite meaningless depending on what you want them for because of the pruning we do on that DB. At minimum the output should make a clear note about that.

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 23]
```diff
@@ -383,63 +410,45 @@ def _get_table_size(config):
...
+                WHERE domain = %s;
+            """, [self.domain])
+            (case_indices,) = cursor.fetchone()
 
         total_cases = resource_model['cases_total']
         if total_cases > 0:
-            resource_model['case_index_ratio'] = total_case_indices/total_cases
+            # total_cases are already multiplied by PARTITIONS_COUNT,
+            # so case_indices needs to also be multiplied in order to cancel out that factor
+            resource_model['case_index_ratio'] = (case_indices * PARTITIONS_COUNT) / total_cases
```
Comment (Charl1996) [inline corehq/apps/reports/management/commands/project_stats_report.py:451]: I was assuming here that the "cases_total" value (which is determined by `CaseES().domain(self.domain).count()`) refers to the _total amount_ of cases, so because the `case_indices` are determined only from 1 shard of the PG DB, I multiplied by PARTITIONS_COUNT in an attempt to estimate the _total amount_ of case indices.

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 22 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py]: @snopoke Mmm...it isn't.

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 20 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py]: Oooh! Good point!

**Comment by Charl1996** [comment]
@snopoke 
> Is it necessary to run queries on all shard databases? Could you not run on 1 and multiply by the number of shard DBs?

Multiplying by the number of shard DBs is a much better approach than querying all shards. I guess missing a couple of rows really does not matter much in the grander scheme of things. 

> The synclog numbers are going to be quite meaningless depending on what you want them for because of the pruning we do on that DB. At minimum the output should make a clear note about that.

I'm in two minds about this...I kept it as is and added a note, but since we're pruning I am starting to feel like removing the synclogs part might be better, since it won't give accurate results in anycase.

**Comment by Charl1996** [comment]
@snopoke 
> I see that we've moved some from using estimate_row_count to doing actual row counts. What was the reason for that change?

The only reason really was because I did a test on staging (ccqa domain) and saw that the estimated value for (cases) rows was 500,000 less than the number counted from ES. This made me prone to using the count from ES, since 500k drift seems like a lot. 

The case indices is much closer though. The estimated value ("The sum of counts from all databases") is 216, whereas the returned value from the SQL statement I have in this PR is 49. Taking the shard value * num_shards, i.e 49 * 5 =  245, it's much closer to the estimate, so I'm happy to change this back to simply using the `estimate_row_count` function.

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 15 above)
Comment (snopoke) [reply corehq/apps/reports/management/commands/project_stats_report.py:72]: bump

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 14 above)
Comment (snopoke) [reply corehq/apps/reports/management/commands/project_stats_report.py:190]: +1

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 24]
```diff
@@ -237,21 +297,24 @@ def _ledgers_per_case(self):
 
         case_types = case_types_result.aggregations.types.keys
 
-        self.stdout.write('\nCase Types with Ledgers')
+        resource_model['case_types_avg_lifespans'] = []
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:300]: why is this part of the 'ledgers' data - it doesn't seem related at all

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 25]
```diff
@@ -208,22 +269,21 @@ def _cases_updated_per_user_per_month(self):
...
-        case_ids = set()
-        ledger_count = 0
         for result in results:
-            case_ids.add(result['case_id'])
-            ledger_count += result['ledger_count']
+            case_ids = result['case_id']
+            ledger_count = result['ledger_count']
 
         if not case_ids:
             self.stdout.write("Domain has no ledgers")
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/project_stats_report.py:289]: should this output be done later?

File: corehq/apps/reports/management/commands/project_stats_report.py [hunk 26]
```diff
@@ -339,8 +400,200 @@ def _get_table_size(config):
...
+                SELECT COUNT(*)
+                FROM form_processor_commcarecaseindexsql
+                WHERE domain = %s;
+            """, [self.domain])
+            (total_case_indices,) = cursor.fetchone()
+
+            total_cases = ResourceModel.get_stat(ResourceModel.CASES_TOTAL)
+            if total_cases > 0:
+                ResourceModel.set_stat(ResourceModel.CASE_INDICES, total_case_indices/total_cases)
+
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/reports/management/commands/project_stats_report.py:456]: I think it can definitely skew things but it really depends on what you want the data for. I guess it would probably be most useful to have both (per case type and average).

**Review by snopoke — APPROVED** [review]
no major comments

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 24 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py]: Oh, it's not. This was part of some change I made and I kinda build it into something that is not represented by "case_types_avg_lifespans" anymore. I changed it now.

File: corehq/apps/reports/management/commands/project_stats_report.py (hunk 15 above)
Comment (Charl1996) [reply corehq/apps/reports/management/commands/project_stats_report.py:70]: > Why not use a dataclass or attrs class?

No particular reason other than the fact that I didn't think of it. However, after having experimented with it I do agree it's an attractive solution.
